src/main/scala/org/graphframes/pattern/patterns.scala

```diff
@@ -48,13 +49,124 @@ private[graphframes] object PatternParser extends RegexParsers {
+        if (edgeNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        vertexNames += name
+      case AnonymousVertex =>  // pass
+    }
+    def addEdge(e: Edge): Unit = e match {
+      case NamedEdge(name, src, dst) =>
+        if (vertexNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        edgeNames += name
+        addVertex(src)
+        addVertex(dst)
+      case AnonymousEdge(src, dst) =>
+        addVertex(src)
+        addVertex(dst)
+    }
+
+    patterns.foreach {
+      case Negation(edge) =>
+        edge match {
+          case NamedEdge(name, src, dst) =>
+            throw new InvalidParseException(s"Motif finding does not support negated named " +
+              s"edges, but the given pattern contained: !($src)-[$name]->($dst)")
+          case AnonymousEdge(AnonymousVertex, AnonymousVertex) =>
+            throw new InvalidParseException(s"Motif finding does not support completely " +
+              s"anonymous negated edges !()-[]->().  Users can check for 0 edges in the graph " +
+              s"using the edges DataFrame.")
+          case e @ AnonymousEdge(_, _) =>
+            addEdge(e)
+        }
+      case AnonymousEdge(AnonymousVertex, AnonymousVertex) =>
+        throw new InvalidParseException(s"Motif finding does not support completely " +
+          s"anonymous edges ()-[]->().  Users can check for the existence of edges in the " +
+          s"graph using the edges DataFrame.")
+      case e @ AnonymousEdge(_, _) =>
+        addEdge(e)
+      case e @ NamedEdge(_, _, _) =>
+        addEdge(e)
+      case AnonymousVertex => // ok
+      case v @ NamedVertex(_) =>
+        addVertex(v)
+    }
+  }
+
+  /**
+   * Split patterns into positive (non-negated) and negated terms, and find all named vertices
+   * in each.
+   * Return the set of named vertices which only appear in negated terms, in sorted order.
+   */
+  private[graphframes]
+  def findNamedVerticesOnlyInNegatedTerms(patterns: Seq[Pattern]): Seq[String] = {
+    val vPos = findNamedElementsInOrder(
+      patterns.filter(p => !p.isInstanceOf[Negation]), includeEdges = false).toSet
+    val vNeg = findNamedElementsInOrder(
+      patterns.filter(p => p.isInstanceOf[Negation]), includeEdges = false).toSet
+    vNeg.diff(vPos).toSeq.sorted
+  }
+
+  /**
+   * Return the set of named vertices (and optionally edges) appearing in the given patterns,
+   * in the order they first appear in the sequence of patterns.
+   * @param includeEdges If true, include named edges in the returned sequence.
+   */
+  private[graphframes]
+  def findNamedElementsInOrder(patterns: Seq[Pattern], includeEdges: Boolean): Seq[String] = {
+    val elementSet = mutable.HashSet.empty[String]
```

Comment:
You may use LinkedHashSet (https://www.scala-lang.org/api/current/scala/collection/mutable/LinkedHashSet.html) to simply the code. It visits elements in the order they were inserted.